File: knapsack.cpp
#include <bits/stdc++.h> 
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
#include<iomanip>
using namespace std; 
 
 
int maximumTotalWeight(vector<int> weights, vector<int> tasks, int p) {

    int n=weights.size();
    vector<vector<int>> opt(n+1,vector<int>(p+1));

    for (int i=0;i<tasks.size();i++){
        tasks[i]*=2;
    }
    
    for (int i=0;i<n+1;i++){
        for(int j=0;j<p+1;j++){
            if (i==0 or j==0){
                opt[i][j]=0;
            }
            else if (j>=tasks[i-1]){
                opt[i][j]=max(opt[i-1][j],weights[i-1]+opt[i-1][j-tasks[i-1]]);
            }
            else{
                opt[i][j]=opt[i-1][j];
            }
        }
    }
    return opt[n][p];
}

int main() 
{ 
    
    int n=4;
    int runtime=15;
    int m=runtime;
    vector<int> length;
    length.push_back(4);
    length.push_back(4);
    length.push_back(6);
    length.push_back(8);
	
    vector<int> weight;
    weight.push_back(2);
    weight.push_back(4);
    weight.push_back(4);
    weight.push_back(5);
    
    for(int i=0;i<n;i++){
    	length[i]*=2;
	}
    
    int x;
    
	cout<<"Enter the number of integers in the dataset"<<"\n";
    cin>>x;
    cout<<"Enter "<<n<<" integers to be sorted"<<"\n";
    

    vector<vector<int> > opt( n+1 , vector<int> (m+1, 0));
	    for (int i=0;i<5;i++){
    	for (int j=0; j<16;j++){
    		cout<<opt[i][j]<<'\t';
		}
		cout<<endl;
	}
	
	cout<<"Solution:"<<endl;
    	    
    for(int i=1;i<=4;i++){
    	for (int j=1;j<=m;j++){
    		if(j>=length[i-1]){
    			opt[i][j]=max(opt[i-1][j],weight[i-1]+opt[i-1][j-length[i-1]]);
			}
			else{
				opt[i][j]=opt[i-1][j];
			}
		}
	}
    
    for (int i=0;i<5;i++){
    	for (int j=0; j<16;j++){
    		cout<<opt[i][j]<<'\t';
		}
		cout<<endl;
	}
    
    return 0;
}

File: knapsack.py
n=int(input("Enter number of items: "))
print("Enter {} weight: ".format(n))
#length=[2,4,4,6]
length=[]
for i in range(0,n):
    length.append(int(input()))
print("Enter {} values: ".format(n))
weight=[]
for i in range(0,n):
    weight.append(int(input()))
#weight=[1,4,6,3]
m=int(input("Max weight: "))

# length=[2,4,4,6]
# weight=[1,4,6,3]
# m=8


opt = [[0] * (m+1) for i in range(n+1)]

# print(opt)
for i in range(0,n+1):
    for j in range (0,m+1):
        if i==0 or j==0:
            opt[i][j]=0
            # print('i=',i,'j=',j)
            # print(opt[0])
        elif j>=length[i-1]:
            # print('i=',i,'j=',j)
            opt[i][j]=max(opt[i-1][j],weight[i-1]+opt[i-1][j-length[i-1]])
            # print(opt[0])
            # print(opt[i])
        else:
            # print('i=',i,'j=',j)
            opt[i][j]=opt[i-1][j]
            # print(opt[0])

# print(length)
print(opt)

print(opt[n][m])


def maximumTotalWeight(weights, tasks, p):
    def knapSack(W, wt, val, n): 
        K = [[0 for x in range(W+1)] for x in range(n+1)] 
    
        # Build table K[][] in bottom up manner 
        for i in range(n+1): 
            for w in range(W+1): 
                if i==0 or w==0: 
                    K[i][w] = 0
                elif wt[i-1] <= w: 
                    K[i][w] = max(val[i-1] + K[i-1][w-wt[i-1]],  K[i-1][w]) 
                else: 
                    K[i][w] = K[i-1][w] 
    
        return K[n][W] 
    for i in range(len(tasks)):
        tasks[i] = 2*tasks[i]
    return(knapSack(p, tasks, weights, len(tasks)))